### PR TITLE
feat: plugin for generating responsive images

### LINKED ISSUE #147: feat: plugin for generating responsive images
[Issuehunt badges](https://issuehunt.io/r/egoist/saber/issues/147)

## Feature request

Plugin that generates responsive images and applies some magic image optimization with optional lazy-loading. Similar to what Gatsby does with `gatsby-image`.

#### What problem does this feature solve?

Slow load times when images are unnecessarily large or not in viewport (with lazy-loading enabled) and developer pain when they have to resize their images manually.

#### What does the proposed API look like?

```yml
# saber-config.yml

plugins:
  - resolve: saber-plugin-image
    sizes:
      large: 1000
      medium: 600
      small: 400
    lazyload: true
    blur: true
```

#### How should this be implemented in your opinion?

Using something like [responsive-loader](https://github.com/herrstucki/responsive-loader) and [vue-lazyload](https://github.com/hilongjw/vue-lazyload)

#### Are you willing to work on this yourself?

Yes.

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/egoist/saber/issues/147).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/egoist/saber/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by krmax44**
**Summary**

Closes #147.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI/CSS related code, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature
- [x] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: packages/saber-plugin-images/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+```js
+import SaberImage from 'saber-plugin-images/SaberImage'
+
+export default ({ Vue }) => {
+  Vue.use(SaberImage, {
+    error: require('./images/error.png')
+  })
+}
+```
```
Comment (egoist) [inline packages/saber-plugin-images/README.md]: why not add a `saber-browser.js` in this plugin?

File: packages/saber-plugin-images/lib/SaberImage.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import VueLazyload from 'vue-lazyload'
+
+export default function(Vue, options = {}) {
+  options = Object.assign({ lazyLoad: true, placeholder: true }, options, {
+    lazyComponent: true
+  })
+
+  Vue.use(VueLazyload, options)
+
+  Vue.component('s-image', {
```
Comment (egoist) [inline packages/saber-plugin-images/lib/SaberImage.js]: I prefer `saber-image`, to align with `saber-link`.

File: packages/saber-plugin-images/package.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "saber-plugin-images",
```
Comment (egoist) [inline packages/saber-plugin-images/package.json]: Maybe `saber-plugin-image`?

File: packages/saber-plugin-images/README.md (hunk 1 above)
Comment (krmax44) [reply packages/saber-plugin-images/README.md]: Oh, I didn't know that was possible :smile:

**Comment by krmax44** [comment]
think that looks pretty good :smile: 
Peek 2019-05-05 18-04

**Comment by MarchWorks** [comment]
How does it look with 0.7s

**Comment by h404bi** [comment]
I like this idea, but for lazyload I prefer using [lozad.js](https://github.com/ApoorvSaxena/lozad.js) combined with a [markdown-it plugin](https://github.com/h404bi/markdown-it-lazy-image), so that I can still use the markdown image syntax while writing articles. :D

**Comment by krmax44** [comment]
@h404bi You will be able to lazy-load Markdown images as well, but if you want to use your own stuff you can disable it.

**Comment by krmax44** [comment]
> I like this idea (the image optimization), but for lazyload I prefer using [lozad.js](https://github.com/ApoorvSaxena/lozad.js) combined with a [markdown-it plugin](https://github.com/h404bi/markdown-it-lazy-image), so that I can still use the markdown image syntax while writing articles. :D

plugin now works with standard markdown images as well :)

@egoist do you think it's ready or is there something still missing?

File: packages/saber-plugin-image/README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
+# saber-plugin-images
+
+Allows the generation of responsive images (`srcset` included), lazy-loading and all that good stuff.
+
+## Install
+
+```bash
+yarn add saber-plugin-images
```
Comment (egoist) [inline packages/saber-plugin-image/README.md]: the readme needs to be updated.

File: packages/saber-plugin-image/lib/index.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+exports.name = ID
+
+exports.apply = (api, options = {}) => {
+  options = Object.assign(
+    {
+      lazyLoad: true,
+      placeholder: true,
+      blendIn: true,
+      markdownImages: false
```
Comment (egoist) [inline packages/saber-plugin-image/lib/index.js]: Maybe it makes sense to enable this by default? If you use this plugin, it's very likely that you want it for markdown images as well 🤔

File: packages/saber-plugin-image/package.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  "main": "lib/index.js",
+  "files": [
+    "lib/index.js",
+    "lib/SaberImage.js",
+    "lib/saber-browser.js"
+  ],
+  "license": "MIT",
+  "dependencies": {
+    "lib": "^4.0.2",
+    "package.json": "^2.0.1",
```
Comment (egoist) [inline packages/saber-plugin-image/package.json]: some weird deps here.

File: packages/saber-plugin-image/lib/index.js (hunk 5 above)
Comment (krmax44) [reply packages/saber-plugin-image/lib/index.js]: Probaby, yeah.

File: packages/saber-plugin-image/lib/SaberImage.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        })
+      }
+
+      const { src, srcSet: srcset } = this.src
+      return h('img', {
+        attrs: { ...$attrs, src, srcset }
+      })
+    }
+  })
+}
```
Comment (egoist) [inline packages/saber-plugin-image/lib/SaberImage.js]: this file won't be transpiled by babel, so either move the code to `saber-browser.js` or let babel include this file.

**Pull request merged**